monitoring/host-proc/analysis.py

```diff
@@ -0,0 +1,616 @@
+                return float(value[:-i]) * self.letters[i][k]
+        return np.nan
+
+
+# Base class for plots and common helpers
+class Plots(metaclass=Singleton):
+    def __init__(self, log_dir, oprefix, jf, to_plot):
+        self.log_dir, self.oprefix = log_dir, oprefix
+        self.df, self.n, self.keys, self.cols = pd.DataFrame(), 0, [], []
+        self.col2title, self.col2units, self.key2nodes = {}, {}, {}
+        self.msg_settling_times, self.msg_injection_times = {}, {}
+        self.grp2idx, self.idx2grp = {}, {}
+        self.fig, self.axes = "", ""
+        self.json_fname, self.G = jf, nx.empty_graph()
+        self.to_plot, self.to_compare = to_plot, []
+
+    # log processing
+    def compute_msg_settling_times(self):
+        ldir = str(self.log_dir)
+        topology_info = topology.load_json(f'{ldir}/{vars.G_TOPOLOGY_FILE_NAME}')
+        topology.load_topics_into_topology(topology_info, f'{ldir}/config/topology_generated/')
+        injected_msgs_dict = log_parser.load_messages(ldir)
+        node_logs, msgs_dict, min_tss, max_tss = analysis.analyze_containers(topology_info, ldir)
+        simulation_time_ms = round((max_tss - min_tss) / 1000000)
+        log.info((f'Simulation started at {min_tss}, ended at {max_tss}. '
+                  f'Effective simulation time was {simulation_time_ms} ms.'))
+        analysis.compute_message_delivery(msgs_dict, injected_msgs_dict)
+        analysis.compute_message_latencies(msgs_dict)
+        self.msg_settling_times = analysis.compute_propagation_times(msgs_dict)
+        self.msg_injection_times = analysis.compute_injection_times(injected_msgs_dict)
+        #print("message propagation_times: ", self.msg_settling_times)
+
+    def get_key(self):
+        return self.df.Key
+
+    def set_keys(self):
+        self.keys = self.df['Key'].unique()
+        self.keys.sort()
+
+    def set_compare(self, lst):
+        self.to_compare = lst
+
+    # extract the maximal complete sample set
+    def remove_incomplete_samples(self, grp, err=''):
+        #if not err:
+        self.df, minRows = self.df[~self.df.isin([err]).any(axis=1)], sys.maxsize
+        for cid in self.df[grp].unique():
+            rows = self.df[self.df[grp] == cid].shape[0]
+            minRows = rows if minRows > rows else minRows
+        self.df = self.df.groupby(grp).head(minRows)
+
+    def plot_msg_settling_times(self):
+        self.set_panel_size(2, 2, False)
+        self.fig.suptitle(f'Settling Time: {len(self.msg_settling_times)} messages')
+        self.fig.supylabel("msecs")
+        pp = PdfPages(f'{self.oprefix}-settling-time.pdf')
+        #axes[0].violinplot([0], showmedians=True)
+        self.axes[0,0].set_xticks([x + 1 for x in range(len(self.keys))])
+        #axes[0].set_xticks(ticks=[x + 1 for x in range(len(self.waku_cids))], labels=self.df["ContainerID"].unique())
+        self.axes[0,0].set_xlabel('TODO: revisit after Jordi added per-container settling times')
+
```

Comment:
What do you mean? (havent seen the plots yet so not sure if you are waiting for me to do something about this)